Downsample/Downsample.cpp

```diff
@@ -0,0 +1,196 @@
+			vis.show(pointCloudDownsampled);
+			vis.resetToFit();
+
+			std::cout << "Running the visualizer. Blocking until the window closes" << std::endl;
+			vis.run();
+		}
+	}
+
+	catch (const std::exception &e)
+	{
+		std::cerr << "Error: " << Zivid::toString(e) << std::endl;
+		return EXIT_FAILURE;
+	}
+}
+
+Eigen::MatrixXf sumline(Eigen::MatrixXf& matrix, int downSamplingFactor)
+{
+	Eigen::MatrixXf reshapedMatrix = matrix.reshaped(downSamplingFactor, (int)matrix.rows() * (int)matrix.cols() / downSamplingFactor);
+	
+	std::function<float(float)> nan_to_zero_wrap = nantozero;
+
+	reshapedMatrix = reshapedMatrix.unaryExpr(nan_to_zero_wrap);
+
+	return reshapedMatrix.colwise().sum().reshaped((int)matrix.rows() / downSamplingFactor, (int)matrix.cols()).transpose();
+}
+
+Eigen::MatrixXf gridsum(Eigen::MatrixXf& r, int downSamplingFactor)
+{
+	return sumline(sumline(r, downSamplingFactor), downSamplingFactor);
+}
+
+float myround(float x)
+{
+	return std::round(x);
+}
+
+float isnanmask(float x)
+{
+	if (std::isnan(x))
+	{
+		return 0;
+	}
+	else
+	{
+		return 1;
+	}
+}
+
+float nantozero(float x)
+{
+	if (std::isnan(x))
+	{
+		return 0;
+	}
+	else
+	{
+		return x;
+	}
+}
+
+Zivid::PointCloud downsample(Zivid::PointCloud& pointCloud, int downSamplingFactor)
+{
+	/*
+	Function for downsampling a Zivid point cloud. The downsampling factor represents the denominator of a fraction that represents the size of the downsampled point cloud relative to the original point cloud, e.g. 2 - one-half,  3 - one-third, 4 one-quarter, etc.
+	*/
+
+	if ((downSamplingFactor % 2 != 0) || (pointCloud.height() % downSamplingFactor) || (pointCloud.width() % downSamplingFactor))
+	{
+		throw std::invalid_argument("Downsampling factor has to have one of the following values: 2, 3, 4, 5, 6.");
+	}
+
+	Eigen::MatrixXf x(pointCloud.height(), pointCloud.width());
+	Eigen::MatrixXf y(pointCloud.height(), pointCloud.width());
+	Eigen::MatrixXf z(pointCloud.height(), pointCloud.width());
+	Eigen::MatrixXi r(pointCloud.height(), pointCloud.width());
+	Eigen::MatrixXi g(pointCloud.height(), pointCloud.width());
+	Eigen::MatrixXi b(pointCloud.height(), pointCloud.width());
+	Eigen::MatrixXf contrast(pointCloud.height(), pointCloud.width());
+
+	for (int i = 0; i < pointCloud.height(); i++)
+	{
+		for (int j = 0; j < pointCloud.width(); j++)
+		{
+			x(i, j) = pointCloud(i, j).x;
+			y(i, j) = pointCloud(i, j).y;
+			z(i, j) = pointCloud(i, j).z;
+			r(i, j) = pointCloud(i, j).red();
+			g(i, j) = pointCloud(i, j).green();
+			b(i, j) = pointCloud(i, j).blue();
+			contrast(i, j) = pointCloud(i, j).contrast;
+		}
+	}
+
+	Eigen::MatrixXf rf = r.template cast<float>();
+	Eigen::MatrixXf gf = g.template cast<float>();
+	Eigen::MatrixXf bf = b.template cast<float>();
+
+	std::function<float(float)> round_wrap = myround;
+
+	Eigen::MatrixXi redDownsampled = ((gridsum(rf, downSamplingFactor) / (downSamplingFactor*downSamplingFactor)).unaryExpr(round_wrap)).template cast<int>();
```

Comment:
Maybe you could extract this line and the `Eigen::MatrixXf rf = r.template cast<float>();` line into a function which takes takes in the `MatrixXi` and the downsampling factor and returns the new `MatrixXi`?

Reply:
done